app/src/tracing.h

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "constants.h"
+
+#include <time.h>
+#include <regex>
+#include <sstream>
+#include <functional>
+#include <ccf/endpoint_context.h>
+#include <ccf/base_endpoint_registry.h>
+
+namespace scitt
+{
+  std::string REQUEST_ID_HEADER = "x-ms-request-id";
+  std::string CLIENT_REQUEST_ID_HEADER = "x-ms-client-request-id";
+
+  std::regex CLIENT_REQUEST_ID_REGEX("^[0-9a-zA-Z-]+");
+
+  thread_local std::string request_id;
+  thread_local std::optional<std::string> client_request_id;
+
+  int diff_timespec_ms(const struct timespec& time0, const struct timespec& time1) {
+    return (time1.tv_sec - time0.tv_sec) * 1000
+        + (time1.tv_nsec - time0.tv_nsec) / 1000000;
+  }
+
+  std::string create_request_id() {
+    std::stringstream stream;
+    stream << std::hex << rand();
+    return stream.str();
+  }
+
+  inline void SCITT_INFO(const std::string& msg)
+  {
+    CCF_APP_INFO("ClientRequestId={} RequestId={} {}",
+        client_request_id.value_or(""),
+        request_id,
+        msg
+        );
+  }
+
+  inline void SCITT_FAIL(const std::string& msg)
+  {
+    CCF_APP_FAIL("ClientRequestId={} RequestId={} {}",
+        client_request_id.value_or(""),
+        request_id,
+        msg
+        );
+  }
+
+  template <typename Fn, typename Ctx>
+  Fn generic_tracing_adapter(
+    Fn fn, const std::function<ccf::ApiResult(::timespec& time)>& get_time)
+  {
+    return [fn, get_time](Ctx& ctx) {
+      request_id = create_request_id();
+      ctx.rpc_ctx->set_response_header(REQUEST_ID_HEADER, request_id);
+
+      client_request_id = ctx.rpc_ctx->get_request_header(CLIENT_REQUEST_ID_HEADER);
+
+      if (client_request_id.has_value())
+      {
+        std::smatch match;
+        if (!std::regex_match(client_request_id.value(), match, CLIENT_REQUEST_ID_REGEX))
```

Comment:
Do we care? We never inspect the value, just forward it around, so it seems fine to allow anything.

Reply:
I added a comment, but yeah, doing some checking is useful as we insert this value directly in log messages and want to avoid misinterpretation of the log, e.g. if it contains whitespaces. It's a bit like SQL injection attacks.